Simplify ActiveModel::Errors#generate_message

DNNX opened the pull request:
### Summary

Besides making `ActiveModel::Errors#generate_message` easier to read, this PR also makes it faster for several reasons:

* We don't eval `@base.class.respond_to?(:i18n_scope)` twice
* We only eval `@base.class.i18n_scope` once
* We don't call `flatten!` because it's not needed anymore
* We don't call `compact!` because all elements are Symbols

### Other Information

```ruby
require 'benchmark/ips'

Benchmark.ips do |x|
  @base = User
  attribute = :email
  type = :invalid

  x.report('before') do
    if @base.class.respond_to?(:i18n_scope)
      defaults = @base.class.lookup_ancestors.map do |klass|
        [ :"#{@base.class.i18n_scope}.errors.models.#{klass.model_name.i18n_key}.attributes.#{attribute}.#{type}",
          :"#{@base.class.i18n_scope}.errors.models.#{klass.model_name.i18n_key}.#{type}" ]
      end
    else
      defaults = []
    end

    defaults << :"#{@base.class.i18n_scope}.errors.messages.#{type}" if @base.class.respond_to?(:i18n_scope)
    defaults << :"errors.attributes.#{attribute}.#{type}"
    defaults << :"errors.messages.#{type}"

    defaults.compact!
    defaults.flatten!
  end

  x.report('after') do
    if @base.class.respond_to?(:i18n_scope)
      i18n_scope = @base.class.i18n_scope
      defaults = @base.class.lookup_ancestors.flat_map do |klass|
        [ :"#{i18n_scope}.errors.models.#{klass.model_name.i18n_key}.attributes.#{attribute}.#{type}",
          :"#{i18n_scope}.errors.models.#{klass.model_name.i18n_key}.#{type}" ]
      end
      defaults << :"#{i18n_scope}.errors.messages.#{type}"
    else
      defaults = []
    end

    defaults << :"errors.attributes.#{attribute}.#{type}"
    defaults << :"errors.messages.#{type}"
  end

  x.compare!
end
__END__
Warming up --------------------------------------
              before    25.251k i/100ms
               after    43.075k i/100ms
Calculating -------------------------------------
              before    276.045k (± 8.2%) i/s -      1.389M in   5.064302s
               after    507.466k (± 6.5%) i/s -      2.541M in   5.028634s

Comparison:
               after:   507465.6 i/s
              before:   276044.9 i/s - 1.84x  slower
```

DNNX committed:
Simplify ActiveModel::Errors#generate_message

Besides making the code easier to read, this commit
also makes it faster:

* We don't eval `@base.class.respond_to?(:i18n_scope)` twice
* We only eval `@base.class.i18n_scope` once
* We don't call `flatten!` because it's not needed anymore
* We don't call `compact` because all elements are Symbols

maclover7 commented:
Hi @DNNX, thank you for your pull request! Can you put together a benchmark to show how this affects methods like `#add` and `#added`?

rafaelfranca commented on activemodel/lib/active_model/errors.rb:
```diff
@@ -398,21 +398,19 @@ def generate_message(attribute, type = :invalid, options = {})
-        defaults = @base.class.lookup_ancestors.map do |klass|
-          [ :"#{@base.class.i18n_scope}.errors.models.#{klass.model_name.i18n_key}.attributes.#{attribute}.#{type}",
-            :"#{@base.class.i18n_scope}.errors.models.#{klass.model_name.i18n_key}.#{type}" ]
+        i18n_scope = @base.class.i18n_scope.to_s
```
Don't need the `to_s` here since it will be interpolated.

DNNX replied:
I did it intentionally because `.to_s` is called in a loop. In order not to convert symbol to string over and over again, I decided to move it up. I'm not attached to it, I can remove it if my argument is not convincing enough.

DNNX commented:
Effect on `#added?`:

```ruby
Benchmark.ips do |x|
  u = User.new
  
  x.report('before') do
    $BEFORE = true
    u.errors.added?(:email, :blank)
  end

  x.report('after') do
    $BEFORE = false
    u.errors.added?(:email, :blank)
  end

  x.compare!
end

__END__
Warming up --------------------------------------
              before   422.000  i/100ms
               after   429.000  i/100ms
Calculating -------------------------------------
              before      4.172k (± 5.5%) i/s -     21.100k in   5.073083s
               after      4.223k (± 6.2%) i/s -     21.450k in   5.098640s

Comparison:
               after:     4222.9 i/s
              before:     4171.9 i/s - same-ish: difference falls within error
```

In order to make this test, I patched my local version of activemodel gem:

```
def generate_message(...)
  if $BEFORE
    <... old implementation ...>
  else
    <... new implementation ...>
  end
end
```

`#add`:

```ruby
Benchmark.ips do |x|
  u = User.new

  x.report('before') do
    $BEFORE = true
    u.errors.add(:email, :blank)
    u.errors.clear
  end

  x.report('after') do
    $BEFORE = false
    u.errors.add(:email, :blank)
    u.errors.clear
  end

  x.compare!
end

__END__
Warming up --------------------------------------
              before   408.000  i/100ms
               after   421.000  i/100ms
Calculating -------------------------------------
              before      3.979k (± 5.0%) i/s -     19.992k in   5.036737s
               after      4.094k (± 5.1%) i/s -     20.629k in   5.052546s

Comparison:
               after:     4093.6 i/s
              before:     3979.1 i/s - same-ish: difference falls within error
```

DNNX commented:
Performance is not my main concern here, I'm more bugged by unneeded `flatten` and `compact`. Performance "improvement" is just a nice side-effect.
